### PR TITLE
feat: Add WhatsApp template messages for onboarding, leaderboard, and DYK with multi-language support

### PR DESCRIPTION
**Pull request by N1kunj1998**
- Add template welcome messages to backend onboarding: Send `onboard_welcome_v2` template after registration for ASHA/ANM users with multi-language support (en/hi/mr/te), using language/type from Excel data
- Convert leaderboard messages to WhatsApp template (`leaderboardv2`) with 9-parameter support
- Convert "Did You Know" messages to WhatsApp template (`did_you_know_v2`) and add Marathi/Telugu language support
- Complete welcome message texts in `THANK_YOU_DICT` (removed truncation placeholders)
- Add error handling and logging for template message delivery

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
**PR Review Summary**

**Overall Impression:**  
The code improvements reflect a strong effort to enhance WhatsApp message handling, especially for template messages, and improve onboarding automation. However, there are several areas for refinement in design clarity, code maintainability, and robustness—particularly around phone number validation, message preparation, and handling multiple languages and user data sources.

---

### Strengths

- **Enhanced multilingual support:** Expanded JSON templates across languages, with proper structure for localization.
- **Robust onboarding flow:** The onboarding script now attempts to personalize welcome messages based on user language and type, with fallback defaults.
- **Whatsapp template logic:** Better handling of message templates, parameters, and API responses; detailed troubleshooting info adds value for debugging.
- **Introduction of utility functions:** Functions like `clean_template_param` and `validate_and_format_phone_number` promote reusability.
- **Testing considerations:** Includes explicit test modes and sample phone validation, enabling safer testing environments.

---

### Critical Observations & Recommendations

#### 1. **Phone Number Validation & Formatting**  
- *Issue:* The validation logic is overly simplistic, only adding '91' for 10-digit Indian numbers, which misses other valid formats and international cases. It also doesn't handle numbers with spaces, symbols, or multiple country codes.
- *Improvement:* Utilize existing libraries like `phonenumbers` for robust validation and formatting. This reduces errors and improves international support.
- *Web Article:* [PhoneNumber validation in Python with phonenumbers](https://github.com/daviddrysdale/python-phonenumbers)

#### 2. **Conditional Complexity & Code Duplication**  
- Multiple sections have repetitive validation, especially in `send_leaderboard_template_messages`. The logic around preparing requests, verifying response types, and printing debug info is verbose and can be centralized into helper functions, reducing cyclomatic complexity.
- *Suggestion:* Modularize response processing and logging—e.g., create a dedicated function to handle API response validation and troubleshooting output.

#### 3. **Handling of Message Parameters & Templates**  
- Creating message params with `clean_template_param` is a good safeguard but lacks comprehensive validation (e.g., length limits, injection defenses).
- The code assumes templates are approved and valid, but these limits (character count, variable placeholders) are not enforced programmatically, risking silent failures.
- *Best Practice:* Before deploying, enforce template limits and check approval statuses via WhatsApp Business API.

#### 4. **Template Payload Assembly & Logging**  
- Extensive debug logs for each request are helpful but clutter the output in production. Consider toggling verbose logs via a debug flag.
- Response handling for `prepare_requests` heavily relies on assumptions about response order and types. Make response processing more explicit.

#### 5. **Cyclomatic Complexity & Readability**  
- The `send_leaderboard_template_messages` function is overly complex (~150+ lines, multiple nested try/excepts). Break it into smaller functions: e.g., `prepare_template_message`, `send_and_log_response`.
- Use consistent style: e.g., avoid multiple nested conditionals, prefer early returns for error states to flatten structure.
- Consider making **test mode** logic clearer: perhaps a dedicated config object or mode parameter instead of inlined booleans.

#### 6. **Data Parsing in Onboarding Script**  
- Handling of `user_location` via `ast.literal_eval()` can be unsafe and error-prone. Better to sanitize or validate JSON before parsing.
- Popping columns and setting defaults is verbose; use pandas `fillna()` or proper dtype conversion for cleaner code.

#### 7. **API Response Handling & Error Management**  
- Responses are sometimes blindly assumed to contain status codes or message responses; add explicit checks and exception handling.
- Log detailed error info on failures, including response content, to assist debugging.

---

### Scores & Final Verdict

**Score: 72/100**  
(Above threshold but with notable room for clarity, robustness, and modularization.)

---

### Final Recommendations

- Adopt `phonenumbers` for phone validation.
- Modularize large, monolithic functions into smaller helpers.
- Incorporate comprehensive response validation and error handling.
- Consolidate logging, especially for debugging API responses.
- Enforce template limits and approval checks before message dispatching.
- Avoid `ast.literal_eval()` for user input; prefer JSON parsing if needed.
- Document assumptions explicitly, e.g., about template variable count and approval status.

**Articles for deeper understanding:**  
- [Python phonenumbers library](https://github.com/daviddrysdale/python-phonenumbers)  
- [Effective Python: Writing Clean and Maintainable Code](https://docs.python.org/3/library/argparse.html)  
- [Handling API responses gracefully](https://realpython.com/api-integration-in-python/)

In sum, these changes are a positive stride but require refinements for production stability, readability, and maintainability.

**Review by a4i-architect — COMMENTED** [review]
**PR Review Summary**

**Overall Assessment:**  
The code introduces comprehensive enhancements, especially in the WhatsApp messaging logic, with a focus on template message handling, robust phone validation, and cleaner user onboarding flows. However, it also adds some complexity, duplicated logic, and areas needing streamlining.

---

### Strengths:
- The inclusion of a `validate_and_format_phone_number` function improves robustness against phone format errors, reducing delivery failures.
- The detailed troubleshooting output in `print_delivery_troubleshooting()` will aid diagnosing WhatsApp message delivery issues.
- Moving from plain text to template-based messaging with parameters enhances scalability and personalization.
- Use of constants and enums (e.g., `MessageTypes`, `LanguageCode`) improves maintainability.
- The onboarding script handles nested data and date fields carefully, prepping data for user registration.

---

### Key Issues & Recommendations:

**1. Overly Complex Logic & High Cyclomatic Complexity**  
- Functions such as `send_leaderboard_template_messages()` are excessively long, handling multiple responsibilities: formatting, validation, request creation, response processing, and print debugging. Break into smaller functions for clarity:
  - `format_leaderboard_params()`
  - `validate_params()`
  - `create_template_request()`
  - `handle_response()`
  
**2. Duplicate/Unnecessary Code**  
- Phone validation logic appears both in the main file and within messaging functions. Centralize validation into one utility function (`validate_and_format_phone_number`) to prevent inconsistencies.
- The code snippets that build `ByoebMessageContext` appear multiple times—abstract this into helper functions to reduce duplication and errors.

**3. Error Handling & Logging**  
- The code relies heavily on `print()` statements. Replace or supplement with a structured logging framework (e.g., Python's `logging`) with appropriate levels (INFO, WARNING, ERROR) for better observability.
- Some exception handling is broad; refine to catch specific exceptions where possible.

**4. Design Pattern & Simplicity**  
- The request processing flow is convoluted, especially when handling template requests among multiple responses. Use helper functions or a builder pattern to streamline:
  - Select and modify template requests
  - Log and verify request payloads before sending
  - Handle response validation separately

**5. Code Duplication & Maintainability**  
- The logic for creating and sending template messages is duplicated for onboarding, leaderboard, and other messaging flows. Consider a generalized `send_template_message()` utility.
- Constants for template names and language codes are well-placed but can be extended for dynamic or externalized configuration.

**6. Response Handling & Status Checking**  
- The response processing for WhatsApp API responses is somewhat fragile: multiple attributes accessed with `hasattr`, and status codes are directly compared. Use a dedicated response handler class or consistent interface to interpret responses.
- The `status` variable extracted as `responses[0].response_status.status` might be insufficient. Instead, parse the entire response object carefully and handle edge cases.

**7. Testing & Simulation**  
- The code demonstrates a good approach with test modes (`TEST_MODE_3_PARAMS`) but could benefit from a formal mock or stub pattern for testing API responses, improving test reliability.

---

### Web Articles for Context & Best Practices
- [Smart API Response Handling and Error Management](https://realpython.com/python-error-handling/)
- [Designing Clean, Modular Code in Python](https://medium.com/swlh/crafting-clean-and-maintainable-python-code-acfd617d6b9f)
- [Best Practices for Phone Number Validation](https://blog.mageplaza.com/validation-for-phone-numbers-in-python/)
- [How to Use Enums for Better Code Readability](https://docs.python.org/3/library/enum.html)

---

### Score & Final Note:
**Score:** **75/100**  
While functional and with noteworthy effort toward robustness and clarity, the code relies on monolithic functions, inconsistent error handling, and verbose debug output that hamper maintainability and clarity. Addressing the points above can bring the score above 85+.

**Minimum threshold met:** Yes, but with scope for significant improvement.

---

**Summary of Recommendations for Next Steps:**  
- Modularize large functions into smaller, single-responsibility helpers.  
- Standardize phone validation and message request creation.  
- Replace print debugging with a proper logging system.  
- Abstract request/response handling into dedicated classes or functions, increasing testability.  
- Document assumptions about data formats and response expectations more clearly.

Implementing these will elevate code quality, ease future maintenance, and reduce bugs in production.

**Comment by N1kunj1998** [comment]
Welcome messages of all languages

image
image
image
image

DYK Messages of all languages

image

Leaderboard messages of all languages
image
image
image

**Review by a4i-architect — COMMENTED** [review]
PR Review Summary:

Overall, the updates aim to enhance robustness, introduce more detailed logging, and improve message formatting for WhatsApp templates, especially around the "leaderboard" and "did you know" features, as well as onboarding messaging. Here are detailed observations:

**Strengths & Positive Aspects:**
- The introduction of phone number validation and formatting functions (e.g., `validate_and_format_phone_number`) are essential for correct WhatsApp message delivery. The logic includes helpful examples and error messages.
- The detailed troubleshooting printouts for delivery issues are a significant improvement for debugging.
- The new structure for sending onboarding messages with dynamic templates, loggers, and error handling enhances robustness.
- The refactor into functions like `send_welcome_message()` improves modularity.
- The onboarding script now better handles variable user data, including potential parsing issues and missing fields.
- The approach for formatting leaderboard data (`format_leaderboard_as_template_parameters`) is flexible, supporting both testing and production modes, and ensures non-empty parameters.

**Areas for Improvement & Concerns:**
1. **Code Duplication & Complexity:**
   - Heavy nested try/except blocks and repeated code snippets (e.g., preparing request, validation, and response handling for each user) could be refactored for clarity and maintainability.
   - A common pattern is to validate, prepare, send, then log. Abstracting this into a reusable function would reduce duplication and aid testing.

2. **Cyclomatic Complexity and Readability:**
   - The core `send_leaderboard_template_messages()` is highly complex, mixing formatting, validation, messaging, and API response handling. Long functions reduce readability and increase error risk.
   - Consider splitting into smaller functions: validation, request creation, response processing, and logging.

3. **Message Template Parameter Handling:**
   - The setup for template parameters (especially in testing mode versus production) moves towards safer defaults but lacks explicit validation for parameter length and content consistency before use.
   - The character limit warning is a good addition but could be automated with a check against actual template body length.

4. **User Data Handling:**
   - The onboarding script relies heavily on `ast.literal_eval()` and manual adjustments, which could be error-prone. Better validation for input consistency, and possibly more robust parsing, would help.
   - The code handling `user_location` assumes it's either a string or a dict; explicitly checking and sanitizing input would increase robustness.

5. **Logging & Debugging:**
   - `print()` statements for debugging are prevalent. Transitioning to a structured logging system with levels (INFO, WARNING, ERROR) improves production readiness.
   - The debug output for WhatsApp request contents is helpful; ensure sensitive data is not logged in production.

6. **Response Handling & Error Checks:**
   - Response status checks rely on `int(responses[0].response_status.status) == StatusCode.SUCCESS.value`; ensure the response object structure matches expectations.
   - Handling API response errors more gracefully and explicitly (e.g., retries, detailed error parsing) would improve resilience.

7. **Overall Design & Patterns:**
   - The current procedural style with deep nesting hampers clarity. Applying object-oriented or functional design principles (e.g., dedicated classes for message preparation, validation, response handling) could streamline this greatly.
   - The "test mode" toggles are scattered; centralizing configuration or flags would simplify toggling behaviors.

8. **Potential Technological Improvements:**
   - Automate template length checks against actual WhatsApp limits.
   - Consider introducing unit tests for core validation and formatting functions.
   - Use of type hints is good, but add return types consistently for better static analysis.

**Score:** 74/100  
While the code now better handles validation, logging, and messaging flows, it remains overly complex and somewhat fragile—mostly due to deep nesting, validation scattered across functions, and lack of abstraction. These issues should be addressed before production deployment.

---

**Web Search Articles & References:**
- [Clean Code – Robert C. Martin (Uncle Bob)](https://en.wikipedia.org/wiki/Clean_Code)
- [Python best practices for complex functions — Real Python](https://realpython.com/lessons/python-clean-code/)
- [Handling API responses effectively — Twilio, WhatsApp API documentation](https://developers.facebook.com/docs/whatsapp/api/message-management/)  
- [Data validation and sanitization in Python — Real Python](https://realpython.com/validation-in-python/)

**Final note:** Prioritize refactoring to reduce complexity and improve maintainability, especially for the messaging functions. Automated tests for validation functions and edge cases are recommended to prevent regressions.

**Review by a4i-architect — COMMENTED** [review]
### Review Summary and Feedback

#### 1. **`background_jobs.py`**
- **Refactor for Async Compatibility**:
  - The change correctly handles sync/async job functions, running sync jobs in an executor to avoid blocking. This is good practice.
  - However, the success/error logging via `_logger.error()` now includes `extra` details which is a good addition.
  - **Suggestion**: Ensure consistent exception logging (e.g., adding traceback info). Consider wrapping the executor call in a dedicated helper to avoid clutter.

- **Function Handling & Routing**:
  - The move to encapsulate job execution (`_execute_job_by_id`) is a good pattern to improve readability.
  - **No issues noted** here.

- **`list_jobs()`**:
  - The logic to fetch scheduled jobs and compile status is clear. Slight optimization: handle missing jobs or potential `None` for `next_run_time` explicitly.  
  - **Suggestion**: Could improve resilience with try-except around scheduler calls.

---

#### 2. **JSON Localization for "Did You Know"—`bot_config.json`**
- **Multilanguage JSON**:
  - Properly added multiparty language support.
  - **No issues**; well-structured multilingual templates.

---

#### 3. **`dyk.csv` Data Files**
- Enhanced with a new last line featuring multilingual entries, correctly formatted.
- No issues.

---

#### 4. **`send_dyk.py`**
- **Major Overhaul with the new approach**:
  - The refactored code is significantly more modular and aligned with best practices:
    - Cleaned template parameters to ensure WhatsApp safety.
    - Uses a consistent approach to prepare and send template messages.
    - Handles invalid number formats with debugging and troubleshooting helpers.
    - Sends messages selectively, and logs responses with detailed status checks.
  - **Design pattern**: Clear separation between data fetching, formatting, and sending.
  - **Overly complicated logic**:
    - The code attempts to handle both test mode and production dynamically. While good, it introduces verbosity, especially with repeated validation and logging. Consider creating utility functions or classes for handling template parameter validation and message state analysis.
  - **Cyclomatic complexity**:
    - Multiple nested `try-except` and if-else blocks; can be modularized further.
    - Response handling, especially for message status validation, is verbose.

- **Suggestions for Improvement**:
  - Centralize phone number validation into a dedicated function.
  - Abstract the response analysis into a helper to reduce clutter.
  - Use consistent, clear logging rather than print statements—especially for debugging.
  - Check for possible missing values (`None`) in parameters before creating messages.

- **Miscellaneous**:
  - The code switches between "test mode" and "production" explicitly. Better to encapsulate this as a parameter object for clarity.
  - The code logs raw API payloads, which is excellent for debugging.
  - Response handling may benefit from more granular checks (e.g., checking `response_status.status == StatusCode.SUCCESS`).

---

#### 5. **`onboarding.py`**
- **Data Preparation Logic**:
  - The handling of Excel data is thorough—parsing dates, nested location fields, missing data.
  - The code attempts to normalize input data to create consistent user records.
  - **Potential issues**:
    - Multiple nested checks and conversions could be simplified using pandas' features more extensively.
    - The code synchronously runs an async function for sending messages, which is acceptable but could be cleaner with an `async main()` pattern for better error handling.

- **Message Sending Logic**:
  - The code now performs safer number and language validation before calling the messaging functions.
  - Skips absent phone numbers gracefully.
  - The new function `send_welcome_messages_to_users()` correctly batches and handles errors.

- **Overall Complexity**:
  - Reasonably well-structured, although some verbose nested logic for location parsing can be refactored.
  - Better error handling around API responses could be added.

---

### General Recommendations
- Emphasize **code modularity**: create utility functions/classes for repetitive tasks like validation, response parsing, parameter formatting.
- Use logging framework (e.g., Python `logging`) throughout instead of `print()` for consistency and better control.
- Reduce cyclomatic complexity by encapsulating nested `try-except` blocks and conditional logic.
- Standardize response validation/error handling from WhatsApp APIs.
- For large static files or structured configs (like JSON, CSV), consider loading dynamically or validating schema.

---

### Web Articles Supporting Feedback
- [Request-Response Handling in Python APIs](https://realpython.com/python-requests/#handling-response-status)
- [Writing Modular Python Code](https://docs.python-guide.org/writing/structure/)
- [Best practices for Asynchronous Python](https://asyncio.readthedocs.io/en/latest/)
- [Logging best practices](https://docs.python.org/3/howto/logging.html)

---

### Final Score: **75/100**
- Sufficient adaptation for async handling, improved structure, deployment-aware features, and detailed logging.
- Still some room for refining complexity, reducing duplication, and clean error handling.

**Acceptance Note**:
- The code demonstrates good progress toward robustness but needs some refactoring to improve readability, reduce cyclomatic complexity, and increase maintainability before full deployment.

---

**Summary**:  
The changes make significant improvements in modularity, async handling, and response validation. Nonetheless, some overly complex nested logic remains, especially in `send_dyk.py`, and error handling could be more robust. The onboarding script also benefits from further abstraction of data normalization. With targeted refactoring focusing on these areas, the system will be more maintainable and reliable.

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 1]
```diff
@@ -299,8 +318,18 @@ async def main(sheet: DykFactSheet, user_types: List[str], batch_size: int) -> N
...
+            continue
+        try:
+            id = str(uuid.UUID(row[guid_col].strip()))  # validate uuids
+        except Exception:
+            run_logger.error("Invalid GUID in %s: %s", SOURCE_PATH.name, row[guid_col])
+            continue
         for lang, lang_col in lang_cols.items():
+            # Guard against missing columns in the row
+            if len(row) <= lang_col:
+                continue
```
Comment (mamuqsit) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: The changes here can be reverted, I believe this was needed to evade the .csv corruption - but we shouldn't have safeguards so that parse errors stay visible (fail fast).

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 2]
```diff
@@ -157,19 +167,28 @@ async def dispatch(dyk_repo: DykRepository, user_repo: UserRepository, sheet: Dy
                 continue
 
             phone_number = user.phone_number_id
-            message = LANG_ENTRIES[record.dyk_lang].template.replace("{message}", sheet[record.dyk_lang][str(record.dyk_id)])
```
Comment (mamuqsit) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py:160]: As we are now using WhatsApp-configured templates, we no longer need LANG_ENTRIES (and the templates defined in bot_config.json) - we can get rid of a lot of the noise now.

<details>
<summary>See diff</summary>

```diff
diff --git a/byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/bot_config.json b/byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/bot_config.json
index 9c25440..f573631 100644
--- a/byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/bot_config.json
+++ b/byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/bot_config.json
@@ -1,34 +1,4 @@
 {
     "path": "dyk.csv",
-    "user_types_to_send": [],
-    "languages": [
-        {
-            "language": "en",
-            "template": [
-                "≡ƒÆí *Did you know?*",
-                "{message}"
-            ]
-        },
-        {
-            "language": "hi",
-            "template": [
-                "≡ƒÆí *αñòαÑìαñ»αñ╛ αñåαñ¬αñòαÑï αñ¬αññαñ╛ αñ╣αÑê?*",
-                "{message}"
-            ]
-        },
-        {
-            "language": "mr",
-            "template": [
-                "≡ƒÆí *αññαÑüαñ«αÑìαñ╣αñ╛αñ▓αñ╛ αñ«αñ╛αñ╣αÑÇαññ αñåαñ╣αÑç αñòαñ╛?*",
-                "{message}"
-            ]
-        },
-        {
-            "language": "te",
-            "template": [
-                "≡ƒÆí *α░«α▒Çα░òα▒ü α░ñα▒åα░▓α▒üα░╕α░╛?*",
-                "{message}"
-            ]
-        }
-    ]
+    "user_types_to_send": []
 }
\ No newline at end of file
diff --git a/byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py b/byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py
index db91013..1fee6f1 100644
--- a/byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py
+++ b/byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py
@@ -18,7 +18,6 @@ from byoeb_core.models.byoeb.message_context import ByoebMessageContext, Message
 from byoeb_core.models.byoeb.user import User
 from byoeb_integrations.channel.whatsapp.meta.async_whatsapp_client import StatusCode
 from datetime import datetime, timezone
-from pydantic import BaseModel, field_validator
 from typing import AsyncIterator, Iterable, List, Optional, Set, Tuple, TypeAlias
 import os
 import re
@@ -26,15 +25,6 @@ import re
 
 DykBatch: TypeAlias = Iterable[Tuple[User, Set[str]]]
 
-class LangEntry(BaseModel):
-    language: LanguageCode
-    template: str  # a template to decorate the message. {message} is the placeholder for the fact.
-
-    @field_validator("template", mode="before")
-    def join_template(cls, v):
-        return "\n".join(v) if isinstance(v, list) else v
-
-
 def clean_template_param(text: str) -> str:
     """Make template parameter safe for WhatsApp: no newlines/tabs, no 4+ spaces."""
     # Replace newlines/tabs with single space
@@ -287,8 +277,6 @@ run_logger = AppInsightsLogHandler.getLogger("dyk_run")
 send_logger = AppInsightsLogHandler.getLogger("dyk_send")
 
 user_types_to_send = bot_config["user_types_to_send"]
-_LANG_ENTRIES = [LangEntry(**e) for e in bot_config["languages"]]
-LANG_ENTRIES = {e.language: e for e in _LANG_ENTRIES}
 N_RETRIES = 5  # number of times to retry dispatch()ing to WhatsApp in the event of failure
 
 SOURCE_PATH = (current_dir / str(bot_config["path"])).resolve()
@@ -302,15 +290,9 @@ with SOURCE_PATH.open(encoding="utf-8") as f:
 
     # fail fast - if these expected cols dont exist, python will bail early
     cols = next(reader)
-    lang_cols = {}
-    for lang in LANG_ENTRIES.values():
-        col = lang.language.value
-        if col not in cols: raise ValueError(f'Column "{col}" does not exist in {SOURCE_PATH.name} - did you forget to create a column for "{col}"?')
-        lang_cols[lang.language] = cols.index(col)
-
+    lang_cols = {lang: cols.index(lang.value) for lang in LanguageCode if lang.value in cols}
     guid_col = cols.index("GUID")
-
-    expected_cols = {"GUID", *[l.language.value for l in LANG_ENTRIES.values()]}
+    expected_cols = {"GUID", *[l.value for l in lang_cols.keys()]}
     unexpected_cols = [c for c in cols if c not in expected_cols]
     if len(unexpected_cols) > 0:
         run_logger.error("Unexpected columns encountered in %s: %s", SOURCE_PATH.name, ", ".join(unexpected_cols))
````
</details>

**Review by mamuqsit — COMMENTED** [review]
Tested send_dyk, message_leaderboard jobs, and onboarding messages through the onboarding script - working across all 4 languages. There are few non-functional changes to polish the PR - I've laid out the steps to approach them.

**Review by a4i-architect — COMMENTED** [review]
### PR Review Summary

#### General Remarks
The changes introduce significant refactoring, especially around the WhatsApp template message handling, with efforts to improve logging, resilience, and message validation. The onboarding script is enhanced for robustness and flexibility. However, there are notable areas where the complexity is unnecessarily high, and code duplication, inconsistency, and verbose logging reduce maintainability. 

---

### Detailed Feedback

#### 1. **Background Jobs: execute_job_function & setup**
- The enhancement to run sync functions in an executor is good practice to prevent blocking. It adds runtime detection, which is prudent.
- The restructuring of `/jobs/{job_id}` into an internal `_execute_job_by_id()` simplifies testing hooks.
- Minor: Repeated use of `job_function.__name__` without existence check (can lead to AttributeError if non-callable).

**Suggestion:** Use `getattr()` with fallback or check `hasattr()` for `__name__` to avoid potential errors.

---

#### 2. **JSON Config & CSV Data Validation**
- Improvement in strict validation (GUID presence, length, cleanup) is essential.
- The nested GUID validation and explicit error handling are strong, but the multiple `len(row)` checks can be cleaned with helper methods for clarity.
- The mix of direct `row[...]` accesses and validation is slightly verbose; consider a helper to streamline validation.

**Suggestion:** Encapsulate row validation and GUID parsing logic.

---

#### 3. **Background Jobs: Did You Know JSON, CSV, and Language Handling**
- Transition from multi-language configuration to a fixed language list (`LANGS`) is sensible, especially for CSV parsing.  
- The code handling CSV parsing is robust, with explicit checks for missing columns and GUID validity, though verbose.
- The `test_mode_3_params` is well implemented, controlling the shape of the leaderboard message parameters.

**Issue:** 
- The logic for padding to 9 parameters and the validation/assertions can be simplified with clearer list initialization.
- Excessive validation/logging can be cleaned up with dedicated functions.

---

#### 4. **Sending Leaderboard Template Messages**
- The core logic for generating and validating template parameters is complex, with multiple assertions and verbose debugging.
- The creation of `ByoebMessageContext` objects with decorated `additional_info` for templates is appropriate.
- The layering of preparing requests, extracting template requests, setting message type, and sending is conceptually correct but over-complicated in places.

**Problems:**
- Overly verbose, nested validation, and debugging statements can be streamlined.
- The assumption that `prepare_requests()` returns both text and template (and how to select them) could be more robust.
- The validation of parameters before sending is duplicated; consolidate into one function.

**Suggestion:** 
- Abstract request creation and validation into helper functions.
- Use a clear pattern: build message, validate request, log, send, and interpret response.

---

#### 5. **Message Status Handling & Troubleshooting**
- The troubleshooting output is helpful but verbose and somewhat redundant.
- The status handling is detailed, which is good, but relies on response attribute assumptions that could cause runtime errors.
- The code for parsing responses and statuses could benefit from a dedicated, more robust parser function.

---

#### 6. **Onboarding Script**
- The code handles reading data, normalizing columns, and merging data gracefully.
- The logic to handle nested dictionaries and parsing `user_location` is commendable, but overly verbose.
- Sending the welcome message is encapsulated well, but the logic for fallback language and validation can be concise with helper methods.

**Issue:**
- The `requests.post()` call and response validation are verbose and can be consolidated.
- The multiple `try/except` blocks and some print statements affect code clarity.

**Suggestion:** 
- Introduce a `sanitize_user_row()` or similar for data extraction.
- Add a `validate_phone_number()` helper for formatting and validation.

---

### Design & Complexity
- The code demonstrates good intentions (robustness, detailed logging), but leads to high cyclomatic complexity, especially in the message sending functions.
- The many nested if-else blocks and repeated validation conditions could be simplified with helper classes/functions.
- The manual debugging logs are verbose; consider a centralized logging and a debug mode toggle.

### Web Articles for Recommendations
- [Simplify complex functions with helper methods](https://refactoring.guru/smells/long-method)
- [Robust JSON handling and response parsing](https://smartling.com/blog/coding-best-practices-in-json)
- [Best practices for async code in Python](https://docs.python.org/3/library/asyncio.html#asyncio-important-concepts)

---

### Final Score: **75/100**

**Strengths:**  
- Improved logging, validation, and exception handling  
- Clear attempt to modularize message creation and sending  
- Better CSV validation and parsing robustness

**Improvements Needed:**  
- **Reduce verbosity and duplicated code** (e.g., request validation, response parsing)  
- **Simplify flow control** with more helper functions and state management  
- **Refactor complex conditional logic**, especially in message send/validation  
- **Add type hints and assert correctness early** to avoid runtime errors  
- **Optimize debug logging** to be more concise and toggle-able

---

**Conclusion:**  
The current changes move the system towards better robustness and correctness but introduce unnecessary complexity that impacts maintainability. With targeted refactoring and simplification, the implementation will be more sustainable and easier to enhance.

**Recommended action:**  
- Address the verbosity and complexity issues in the critical functions, especially message generation and response handling.  
- Implement helper utilities for validations and request handling.

Once refactored accordingly, the codebase will be significantly improved.

**Comment by mamuqsit** [comment]
Looks good! Quick question - was the "More soon." added deliberately to DYKs? It adds a minor inconvenience where WA truncates a message with a 'Read more' on my device.
image

image

**Review by a4i-architect — COMMENTED** [review]
PR Review Summary:

Overall, the code demonstrates significant improvements towards robustness and better structuring, especially in the messaging and onboarding workflows. Key modifications include improved async handling, structured template message preparation, thorough phone number validation, and clearer debugging outputs.

Strengths:
- The transition from plain text to template-based WhatsApp messages is well-executed, with explicit checks and detailed debug logs.
- Validation functions for phone formatting and message payloads are thoughtful, reducing the risk of runtime errors.
- The onboarding script now handles nested and malformed data more gracefully, with fallbacks.
- Sending logic now emphasizes error handling and debugging, making troubleshooting easier.

Areas for Improvement:
1. **Code Duplication & Overcomplication**: 
   - The `send_leaderboard_template_messages` function is overly verbose—there's repeated logging, redundant message building, and complex message/timing handling that could be encapsulated into smaller helpers.
   - The main send logic mixes multiple responsibilities: validation, message building, WhatsApp API interaction, and logging. Extracting these into smaller, testable units would improve clarity and maintainability.

2. **Type Safety & Consistency**:
   - The validation of `template_parameters` relies heavily on runtime assertions and manual checks. Introducing pydantic models or strict type annotations for message payloads would increase robustness.
   - Handling of `requests` from `prepare_requests()` is fragile—assuming the presence of `"type"` keys without checks could cause runtime errors.

3. **Cyclomatic Complexity & Control Flow**:
   - The flow inside the `try-except` loop sending messages is complex, with nested conditionals for formatting, validation, response handling, and debug output. Simplifying with dedicated functions and early exits would be beneficial.
   - The `main()` function includes commented-out code for different modes. A configuration-driven approach (e.g., a "mode" variable) would make switching modes clearer.

4. **Hardcoded & Magic Values**:
   - Test phone numbers, sleep durations, character limits, and template names are hardcoded or embedded in logs. Moving configurations into constants or environment variables enhances flexibility.
   
5. **Web Search Insights**:
   - Articles like [Best Practices for Building Chatbot Messaging](https://chatbotsmagazine.com/best-practices-for-building-chatbot-messaging-12258f3391d8) emphasize validation and minimizing complexity in message workflows.
   - [WhatsApp Business Template Management](https://developers.facebook.com/docs/whatsapp/message-templates/) highlights the importance of template approval and character limits, directly relevant to the character limit warning present.

**Score: 78/100**  
(Respectable improvements, but areas like complexity, code clarity, and type safety need focused refactoring before this is production-ready.)

**Main Recommendations**:
- Modularize large functions, particularly message composition and validation.
- Enforce consistent data typing, possibly via pydantic models.
- Externalize configuration values such as delays and template names.
- Simplify control flow, favor early returns and helpers.
- Consider adding unit tests around phone validation and message building.

This approach aligns with best practices in chatbot development and API integrations, reducing bugs and improving maintainability.

**References**:
- [Chatbot Messaging Best Practices](https://chatbotsmagazine.com/best-practices-for-building-chatbot-messaging-12258f3391d8)
- [WhatsApp Business API Template Guidelines](https://developers.facebook.com/docs/whatsapp/message-templates/)
- [Writing Readable and Maintainable Async Python](https://realpython.com/async-python/)

This review recommends targeted refactoring, especially reducing cyclomatic complexity and increasing modularity.

**Review by mamuqsit — APPROVED** [review]
